sklearn/tree/_splitter.pyx

```diff
@@ -181,8 +188,27 @@ cdef class Splitter:
 
         self.sample_weight = sample_weight
 
-    cdef void node_reset(self, SIZE_t start, SIZE_t end,
-                         double* weighted_n_node_samples) nogil:
+        n_samples_total[0] = j
+        weighted_n_samples_total[0] = weighted_n_samples
+
+        self.start = 0
+        self.end = j
+
+        self.criterion.init(self.y,
+                            self.y_stride,
+                            self.sample_weight,
+                            self.weighted_n_samples,
+                            self.samples,
+                            self.start,
+                            self.end)
+
+        sum_total[0] = self.criterion.sum_total
+        sq_sum_total[0] = self.criterion.sq_sum_total
+        impurity[0] = self.criterion.node_impurity()
+
+    cdef void node_reset(self, SIZE_t start, SIZE_t end, 
+                         double weighted_n_node_samples, 
+                         double* sum_total, double sq_sum_total) nogil:
```

Comment:
For the story, `node_reset` exists because my intent was for `Builder` objects to only interact with `Splitter` objects, and for `Splitter` objects to only interact with `Criterion` objects.